secrets/manager.go

```diff
@@ -0,0 +1,216 @@
+// Copyright 2023 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package secrets
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"gopkg.in/yaml.v2"
+)
+
+var (
+	failedSecretConfigs = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Name: "prometheus_sp_failed_secret_configs",
+			Help: "Current number of secret configurations that failed to load.",
+		},
+	)
+	secretsTotal = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Name: "prometheus_sp_secrets_total",
+			Help: "Current number of secrets.",
+		},
+	)
+)
+
+func init() {
+	prometheus.MustRegister(failedSecretConfigs)
```

Comment:
We try to avoid global instrumentation if possible and try to register those when NewProviderManager is actually instantiated (even if only one is allowed), but not a blocker.

Reply:
Are there any examples where we _don't_ do global instrumentation in upstream Prometheus? All current service provider instrumentation is enabled by default. I couldn't find any place where we registered instrumentation on instantiation.

Reply:
For example here: https://github.com/prometheus/prometheus/blob/e2b9cfeeeb1efbe299c386226622a5064f23f377/storage/remote/write.go#L79